src/vec_poly.h

```diff
@@ -0,0 +1,234 @@
+ * \f}
+ *
+ * Note, derivate of \f$X^n\f$ is defined as sum of \f$n\f$ polynomials
+ * \f$X^{n-1}\f$, not a multiplication of \f$n\f$ and \f$X^{n-1}\f$. It can be
+ * expressed as below:
+ * \f{eqnarray*}{
+ *  derivative(X^n) &= (1 + 1 + ... + 1) X^{n-1}
+ *                  &= (n % p) * X^{n-1}
+ * \f{eqnarray*}{
+ * where \f$p\f$ is the characteristic of the field.
+ *
+ * @return
+ */
+template <typename T>
+void Poly<T>::derivative()
+{
+    for (int deg = 1; deg <= degree; ++deg) {
+        buf[deg - 1] = field->mul(deg % field_characteristic, buf[deg]);
+    }
+    buf[degree] = 0;
+    degree--;
+}
+
+template <typename T>
+T Poly<T>::eval(T x)
+{
+    int i = degree;
+    T result = buf[i];
+
+    while (i >= 1) {
+        result = field->add(field->mul(result, x), buf[i - 1]);
+        i--;
+    }
+    return result;
+}
+
+/** Multiply to a polynomial \f$b\f$ with a limited degree of result
+ *
+ * Degree of result polynomial is limited by a given value, i.e. elements of
+ * degree greater than this value are ignored. It reduces operations in cases
+ * the degree limit is smaller than degree of input polynomials
+ *
+ * @param b - polynomial to multiply
+ * @param deg_out - degree of result
+ */
+template <typename T>
+void Poly<T>::mul(Poly<T>* b, int deg_out)
+{
+    assert(deg_out >= 0 && deg_out < this->n);
+
+    for (int deg = deg_out; deg >= 0; deg--) {
+        T val = 0;
+        for (int deg_a = 0; deg_a <= deg && deg_a <= degree; deg_a++) {
+            const T val_a = buf[deg_a];
+            if (val_a > 0) {
+                const int deg_b = deg - deg_a;
+                const T val_b = b->get(deg_b);
+                if (val_b > 0)
+                    val = field->add(val, field->mul(val_a, val_b));
+            }
+        }
+        buf[deg] = val;
+    }
+}
+
+/** Multiply to polynomial \f$(X + coef)\f$
+ *
+ * \f{eqnarray*}{
+ * P(X) = &a_0 + a_1 X + ... + a_n X^n \\
+ * Q(X) = &coef + X \\
+ * R(X) = &P(X) \times Q(X) \\
+ *      = &a_0 * coef + a_1 * coef X + ... + a_n X^n + \\
+ *        &a_0 X + a_1 X^2 + ... + a_{n-1} X^n + a_n X^{n+1} \\
+ *      = &a_0 * coef + sum_{i=1}^{n}(a_{i-1} + a_i * coef) X^i + a_n X^{n+1}
+ * \f}
+ *
+ * @param coef - coef of the multiplied polynomial (X + coef)
+ */
+template <typename T>
+void Poly<T>::mul_to_x_plus_coef(T coef)
+{
+    assert(field->check(coef));
+
+    const T val = buf[degree];
+    for (int deg = degree; deg > 0; deg--) {
+        buf[deg] = field->add(buf[deg - 1], field->mul(buf[deg], coef));
+    }
+    buf[0] = field->mul(buf[0], coef);
+    set(degree + 1, val);
+}
+
+template <typename T>
+void Poly<T>::neg()
+{
+    for (int i = 0; i <= degree; i++)
+        buf[i] = field->neg(buf[i]);
+}
+
+template <typename T>
+void Poly<T>::zero()
+{
+    std::memset(buf, 0, this->n * sizeof(*buf));
+}
+
+template <typename T>
+void Poly<T>::dump()
+{
+    for (int i = 0; i <= degree; i++)
+        std::cout << buf[i] << " x^" << i << " + ";
```

Comment:
This will result in a weird output, no? Somehting like "ax^0 + bx^1 + " (trailing +).

I think this could be fixed as follow:

```C++
    for (int i = 0; i <= degree; i++) {
        if (i) {
            std::cout << " + ";
        }
        std::cout << buf[i] << " x^" << i;
    }
```